### DISCUSSION TITLE
Hajk 3.12

### DISCUSSION THREAD

**Discussion by Hallbergs** [hajkmap/Hajk#1350] [General]
_Edit: Updated to reflect the changes in Release Candidate 2_

[211 commits](https://github.com/hajkmap/Hajk/compare/v3.11.0...master) and changes in 161 files since the 3.11 release – here's Hajk 3.12! This release was brought to you by @Hallbergs  and @jacobwod . Please see the list of commits for all contributors.

This release brings a lot of improvements both in Client and the NodeJS backend. 

There are of course upgrades to the latest dependencies that ensure that your Hajk setup is up-to-date when it comes to **the latest security patches**. 🔒  

Continue reading for a details list of changes! 🚀 

# Breaking changes
As of #1332, The NodeJS backend requires **Node v16.10** or greater. However, it is recommended to always use the latest NodeJS LTS, which at time of writing is [v18](https://github.com/nodejs/release#release-schedule).

# Changes since Hajk 3.12 Release Candidate 1
A bug was discovered regarding the timing of proxy initialisation in the NodeJS backed. This could lead to the FME proxy not being initiated as intended. This problem is fixed in 45565ae861afbe8cb6356f88f5083439c4c9288a which is included in RC2. 

# Major new features
## Live app state in URL's hash
#1252 adds a new major function: everything the user does in Hajk is kept in a portion of the URL called _query string_ (sometimes called _URL hash_, due to it's leading character being a `#`). 

### What to expect
As users interacts with the map, URL changes:

https://github.com/hajkmap/Hajk/assets/110222/275e6532-a8d6-40a4-9c08-d35ca61f55ea

### Benefits
The main benefits are:
- if browser is set to restore tabs on start, Hajk will remember its latest state (visible layers, location, zoom level, active search string, and more)
- user has the option to easily add bookmarks to "favourite" maps, kind of creating own _theme maps_
- it is super-easy to share whatever is visible with e.g. a colleague, without invoking the Anchor plugin. Just copy the URL and send.

### How to activate
The goal is to make this new functionality the new default way that Hajk works (perhaps even require it). But currently, we will phase-in it. **You, as an admin, must manually enable it on maps.**

You can find the switch in Admin UI:
![image](https://github.com/hajkmap/Hajk/assets/110222/3d3fbb52-4f70-43c9-b0f3-7ad77c8df592)

Another option is to test it without any changes in the config. To do that, just **add the `appStateInHash` parameter to your existing URL in the search query portion** (the part after a `?`, e.g. `http://localhost/?m=map_1&enableAppStateInHash`).

**PLEASE NOTE:** The two methods of enabling this functionality are not related. **If you enable the setting using Admin UI, you don't need to add anything to the URL.**

## Additions to the TimeSlider
The TimeSlider plugin now allows for quarterly steps. We've also added functionality allowing the user to step one step at the time.
![image](https://github.com/hajkmap/Hajk/assets/43059093/2a8d30de-f7ff-43ba-bbae-7ac95a9470bd)

## Additions to the Infoclick window
It is now possible to use custom URIs in the InfoClick-window.
![image](https://github.com/hajkmap/Hajk/assets/43059093/d37c3dd1-ccd7-4fc9-8666-0f24e7173d87)

## Configurable drawer icon
The `DocumentHandler` drawer button icon can now be configured in the admin interface.
![image](https://github.com/hajkmap/Hajk/assets/43059093/72a5f799-1417-4009-bf7c-356b653f10d8)

## Bookmarks plugin cookie verification
Before storing data in `localStorage`, the plugin now verifies if it has the user's permission to use cookies.
![image](https://github.com/hajkmap/Hajk/assets/43059093/830cb801-d954-4345-8ed1-c51e0bd33f2a)

## Zoom level indicators update
Both checkboxes and `Snackbars` will now appear consistently for both individual `layers` and `group layers` when activated outside their specified zoom level.
![image](https://github.com/hajkmap/Hajk/assets/43059093/c61db61d-45f8-4c9c-b71a-a37dbcebe2bd)

# Changelog
For closed issues marked with the 3.12 milestone, [see this](https://github.com/hajkmap/Hajk/milestone/13?closed=1).

For a complete list of commits, [see this](https://github.com/hajkmap/Hajk/compare/v3.11.0...master).

# New features and bug fixes

#1338, #1330, #1329, #1326, #1323, #1319, #1309, #1289, #1269, #1280, #1268, #1267, #1266, #1265, #1264, #1258, #1252, #1243, #1219, #1202, #1342, #1344, #1336, #1328, #1334, #1314, #997, #1308, #1302, #1301, #1294, #1305, #1286, #1282, #1278, #1277, #1274, #1324, #1292, #1272

# Downloads
There are multiple ways to get the latest Hajk up and running, including Docker container.

## Pre-compiled releases
As usual, we have prepared **three different** versions of the pre-compiled solution (scroll down for download links):
1. [Hajk 3.12 RC 2 NodeJS edition](https://github.com/hajkmap/Hajk/releases/download/v3.12.0-rc.2/hajk-v3.12.0-rc.2-nodejs.zip)
1. [Hajk 3.12 RC 2 Simple edition](https://github.com/hajkmap/Hajk/releases/download/v3.12.0-rc.2/hajk-v3.12.0-rc.2-simple.zip)
1. [Hajk 3.12 RC 2 .NET edition](https://github.com/hajkmap/Hajk/releases/download/v3.12.0-rc.2/hajk-v3.12.0-rc.2-dotnet.zip)

## Docker container
The container is available on Docker Hub: https://hub.docker.com/r/hallbergs/hajk

# Installation
Refer to our [installation guide](https://github.com/hajkmap/Hajk/wiki/Installation-guide-%28for-pre-packaged-releases%29).

------

# What's Changed
* Feature/1219 configurable drawer button by @linusfj in https://github.com/hajkmap/Hajk/pull/1255
* Added functionality for adding frame around image in admin by @OlofSvahnVbg in https://github.com/hajkmap/Hajk/pull/1251
* Feature/1099 infoclick parse xml by @linusfj in https://github.com/hajkmap/Hajk/pull/1256
* Fix/1258 extend feature info support by @Hallbergs in https://github.com/hajkmap/Hajk/pull/1262
* Feature/1269 props as filter args by @jesade-vbg in https://github.com/hajkmap/Hajk/pull/1271
* Feature/1252 control hajk programatically by @jacobwod in https://github.com/hajkmap/Hajk/pull/1272
* Feature/1267 refactor bookmarks plugin by @linusfj in https://github.com/hajkmap/Hajk/pull/1270
* Feature/1264 bookmarks cookies verification by @linusfj in https://github.com/hajkmap/Hajk/pull/1283
* Feature/1268 accordion blockqoute by @OlofSvahnVbg in https://github.com/hajkmap/Hajk/pull/1279
* Feature/1243 new measure tool by @jesade-vbg in https://github.com/hajkmap/Hajk/pull/1292
* Feature/1280 print scale bar update by @OlofSvahnVbg in https://github.com/hajkmap/Hajk/pull/1293
* Fix that closes #1314 -dh2.0-large-images-breaks-view by @jesade-vbg in https://github.com/hajkmap/Hajk/pull/1315
* Fix that closes #997. Its now possible to specify image (and other st… by @jesade-vbg in https://github.com/hajkmap/Hajk/pull/1313
* feature/1309: API version separation in NodeJS backend by @jacobwod in https://github.com/hajkmap/Hajk/pull/1318
* Reinstated cacheBuster to new-client by @jesade-vbg in https://github.com/hajkmap/Hajk/pull/1324
* Feature/PrintAttachmentsMerge by @OlofSvahnVbg in https://github.com/hajkmap/Hajk/pull/1322
* Getting custom fonts to work in DocumentHandler Print. by @jesade-vbg in https://github.com/hajkmap/Hajk/pull/1327
* Feature/1329 timeslider additions by @Hallbergs in https://github.com/hajkmap/Hajk/pull/1331
* bug fix - Feature/print attachments merge by @jesade-vbg in https://github.com/hajkmap/Hajk/pull/1333
* Make the Backend fully ESM compatible, remove Babel by @jacobwod in https://github.com/hajkmap/Hajk/pull/1332
* Potential fix for #1320: by @jacobwod in https://github.com/hajkmap/Hajk/pull/1337
* Fix/1291 snackbar group layers by @linusfj in https://github.com/hajkmap/Hajk/pull/1304
* Feature/1319 open desktop app in infoclick by @Hallbergs in https://github.com/hajkmap/Hajk/pull/1341
* Feature/1338 anchor bug fix by @OlofSvahnVbg in https://github.com/hajkmap/Hajk/pull/1339
* Feature/orebro tsm 2 by @Sunix71 in https://github.com/hajkmap/Hajk/pull/1281

**Full Changelog**: https://github.com/hajkmap/Hajk/compare/v3.11.0...v3.12.0-rc.2

<hr /><em>This discussion was created from the release <a href='https://github.com/hajkmap/Hajk/releases/tag/v3.12.0-rc.2'>Hajk 3.12.0 Release Candidate 2</a>.</em>

**Comment by Hallbergs** [comment]
A major bug has been detected in the NodeJS-backend, killing FME-server-proxies etc. Fix coming up!

**Comment by Hallbergs** [comment]
A new release with updated code can be found here: https://github.com/hajkmap/Hajk/releases/tag/v3.12.0-rc.2

**Comment by elizabethgbg** [comment]
Great work!
We’ve tested some of the cases and these are our findings:

**DocumentHandler 2.0 - Large images with width and height set breaks view. #1314**
We have a lot of images which we haven't specified the sizes for (which is said to be ok). When printing, those images are no longer handled correctly.
Large white areas appear above and below the image when printing. See examples below where the first image is set without size and where the second image is sized.
![Bild1](https://github.com/hajkmap/Hajk/assets/120728926/8f8f2c9a-1992-4f96-a889-183a876df1c2)

![Bild2](https://github.com/hajkmap/Hajk/assets/120728926/4453ae6a-e3cf-40d7-a55e-2d7df01d3cfb)


**DocumentHandler - Print does not use customFont #1326**
I'm not sure if the bug comes from this case or from some other change made to the print function in the document handler. The font loads as it should, but problems arise when we print large amounts of text.
All our body text disappears when printing our entire översiktsplan. Only italic text and some layout are included. If we print a single page, it works fine, then all the text is included. It's like it doesn't have time to render everything correctly.
See examples below from appearance in the client and how it looks when printing to pdf.
![Bild3](https://github.com/hajkmap/Hajk/assets/120728926/553ada69-5f7a-4dea-9ac8-1536dd7f0526)

![Bild4](https://github.com/hajkmap/Hajk/assets/120728926/796c5b2d-a0f7-4d00-a62f-e3b15cd43e94)


**Print Scalebar Update #1280**
We don’t find the scale bar to be optimal in all scales at the moment. For example, at a scale of 1:7 500, the scale bar is 375m, which is not an even number. At smaller scales, e.g. at the scale of 1:400, the scale bar becomes very short - is it possible to tweak the component a bit?

**New Measure tool - step 1 #1243**
Is it possible to ad a tooltip to the “point” in this tool, that describes that the function shows coordinates?

**Make it possible to link to other service/main site from Hajk #1336**
.NET-backend looses the specified settings when we save something again in admin.

**Zoom Control: Not possible to disable zoom animation due to ignored zoomDuration #1334**
.NET-backend cannot write the value to the config file.

**Reply by emmafrygne-vbg** [reply]
Hi! I have also tested:
DocumentHandler 2.0 - Large images with width and height set breaks view. [#1314](https://github.com/hajkmap/Hajk/issues/1314)
DocumentHandler - Print does not use customFont [#1326](https://github.com/hajkmap/Hajk/issues/1326)

I don´t know if its any help but I don't experience the problems that Elizabeth wrote about.

**Reply by elizabethgbg** [reply]
**Regarding [#1314](https://github.com/hajkmap/Hajk/issues/1314)**
We have been in contact with Varberg and found that their images have width and height specified, so we suspect that the bug can exist after all.

Please try if it's possible for you to reproduce the bug:
In the document, manually remove width and height from an image and then print the document.
You can also try a document from us, where the images don't have width and height specified:
[Document for test.zip](https://github.com/hajkmap/Hajk/files/12549603/Document.for.test.zip)

Height and width did not have to be specified when we created our översiktsplan. We therefore have a lot of images where this is not specified which now causes problems when we print documents.

**Regarding [#1326](https://github.com/hajkmap/Hajk/issues/1326)**
We found that the usage of the µ symbol in our texts was the reason we were experiencing problems. We replaced the symbol with the HTML entity code which solved the problem.

**Comment by OlofSvahnVbg** [comment]
Regarding **Print Scalebar Update #1280**

It seems the issues you're describing can be fixed by tweaking inputs for "Skalor" and "Skalmeter" in admin. For example, in the image below the value for scale 1:500 is set to 50 meters:
![image](https://github.com/hajkmap/Hajk/assets/111860819/311ea135-4044-4519-bd03-1fa5ef37b1db)

This makes the scalebar look like this:
![image](https://github.com/hajkmap/Hajk/assets/111860819/f1e3851e-4f84-4ad5-89d6-7505f12129f2)

However, if I change the value for "Skalmeter" to say, 10, the scalebar looks lite this:
![image](https://github.com/hajkmap/Hajk/assets/111860819/6675e463-40ac-42e5-a24b-b931562d7c8c)

You can change these values as you please.
Hopefully, this solves your issues, if not, please get back to me :)
/ Olof Svahn.

**Reply by elizabethgbg** [reply]
Now it works, thanks!

Is it possible to find these kinds of instructions, on how to activate/use new functions, somewhere when we do RC-tests? If not – is it possible to make short instructions available for testers in the future? It would help a lot! 

Regarding the scale bar:
I saw that the 0 and 100 are very close to each other with this combination: _scale 25,000_ and _scale meter 1000_, see image below. Is it possible to change this?
![Skärmklipp](https://github.com/hajkmap/Hajk/assets/120728926/24a8807e-fd7f-4200-aecf-a8ace5b06a0c)

**Comment by OlofSvahnVbg** [comment]
Im not sure on about the instructions, but I dont see why not. Sounds like
a good idea. Will check on it.

So, the short answer regarding the 0 100 being too close, is change the
scalebar "Skalmeter" to something bigger. Like this:
[image: image.png]
Fixing this is somewhat problematic, and frankly, maybe not worth the time.
Seeing as how the user can easily change the values themselves.

Might be a fix in the future.

Den ons 23 aug. 2023 kl 16:04 skrev Elizabeth Barlow <
***@***.***>:

**Comment by jesade-vbg** [comment]
@elizabethgbg
Regarding:
"New Measure tool - step 1 https://github.com/hajkmap/Hajk/issues/1243
Is it possible to ad a tooltip to the “point” in this tool, that describes that the function shows coordinates?"

It will be changed in next release 3.13
![image](https://github.com/hajkmap/Hajk/assets/73874338/35ee3c9b-f703-4eb7-ba2d-c7ef16781869)

**Comment by jesade-vbg** [comment]
#1314 Has been reverted. It had to much impact to be fixed in 3.12 and prevented the 3.12 release. It will probably be revisited in 3.13.
@elizabethgbg @lina-sbk-gbg

**Comment by Hallbergs** [comment]
3.12 has now been released! Check it out [here!](https://github.com/hajkmap/Hajk/releases/tag/v3.12.0)
